### DISCUSSION TITLE
How to force invoke a plugin

### DISCUSSION THREAD

**Discussion by Skyquek** [microsoft/semantic-kernel#9003] [Q&A] [answered]
**[Issue]**
I want to invoke a plugin that can help me to generate graph in JSON format. However, I am facing problem where sometimes the plugin is invoked, sometimes it don't (hallucinate). I am using `AutoInvokeKernelFunction`.

I am checking here what should I do in order to make sure my plugins always get invoked?

Thanks!

**Comment by moonbox3** [comment, marked as answer]
You can use the `Required` FunctionChoiceBehavior (or its equivalent depending upon the AI service). This tells the model it must make a tool call every time. The `Required` settings is a bit tricky, though. You will need to figure out how many "auto invoke attempts" to configure, as each time you call the LLM it will only respond with a tool call (no need to waste tokens continuously calling the same function).

**Reply by HuskyDanny** [reply]
Hi @moonbox3 , I am confused with the part "as each time you call the LLM it will only respond with a tool call", does this mean with Required, the llm will not return any content but only the tool call? 

So this behavior is different from Auto? My scenario is to have the chat function always call one of the functions, so setting the functionChoiceBehavior as required.

**Reply by moonbox3** [reply]
Hello @HuskyDanny, yes, that is correct. If your `FunctionChoiceBehavior` is configured as `Required`, the model _will _only_ make a tool call. You would need to have logic to control if you want to change the function choice behavior so the model can use any FunctionResultContent to form a final response, at some point.

Do all keep in mind that OpenAI models allow one to send multiple `Required` tools; however, Azure OpenAI only allows for one `Required` tool at a time. (I need to check if/when they will support multiple as OpenAI does.)

**Reply by yanshinian** [reply]
Is there a parameter to control this behavior? Currently, after setting 'Required', it returns the result of the tools call. I want to locally invoke the plugin and then use the result from that invocation to make another call to the llm.Just like when I use 'auto', it automatically calls the LLM for the second time.

**Reply by moonbox3** [reply]
@yanshinian, when configuring `FunctionChoiceBehavior.Required()`, the model always calls a tool to satisfy the query/ask. Depending on your `maximum_auto_invoke_attempts` configured as part of `FunctionChoiceBehavior` (the default being 1 if not configured), the model will continuously call a tool. Since it's configured to 1 call by default, you will get a tool call, the function is invoked, and then we get kicked out of the auto invocation loop. The next call to the model doesn't advertise any tools, so the model responds in natural language based on the function result from the previous invocation. This is assuming you don't toggle `auto_invoke=False`. 

If you do configure it like:

```python
function_choice_behavior=FunctionChoiceBehavior.Required(
    auto_invoke=False,
    filters={"included_functions": ["time-time", "time-date"]},
),
```

You will get back a `FunctionResult` (if invoked via `kernel.invoke(...)`) or a `ChatMessageContent` if calling the AI service directly, that contains an item of type `FunctionCallContent`. You are then responsible for calling your function or calling `kernel.invoke_function_call(...)` and then returning the `FunctionResultContent` as part of the chat history as the model requires that type to follow `FunctionCallContent`. You would need to then disable FunctionChoiceBehavior, or configure it to `.Auto()` if you want the model to potentially call a tool or form a natural language response.

**Reply by moonbox3** [reply]
@yanshinian we have this sample showing the behavior with `Required` and `auto_invoke=True`. You can easily toggle it to false to see it print the tool calls.

https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/auto_function_calling/function_calling_with_required_type.py
